File: src/main/java/com/example/cqrs/command/entity/event/MoneyDepositedEventEntity.java
package com.example.cqrs.command.entity.event;

import com.example.cqrs.command.entity.event.metadata.EventMetadata;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("MoneyDepositedEvent")
@Entity
public class MoneyDepositedEventEntity extends AbstractAccountEventEntity {

    public MoneyDepositedEventEntity(String accountId, LocalDateTime eventDate,
                                     double amount, EventMetadata metadata) {
        super(accountId, eventDate, amount, metadata);
    }

    // factory method
    public static MoneyDepositedEventEntity of(String accountId, LocalDateTime now, double amount, EventMetadata eventMetadata) {
        return new MoneyDepositedEventEntity(accountId, now, amount, eventMetadata);
    }

}

File: src/main/java/com/example/cqrs/command/entity/event/MoneyDepositedEventEntity.kt
package com.example.cqrs.command.entity.event

import com.example.cqrs.command.entity.event.event.AbstractAccountEventEntity
import com.example.cqrs.command.entity.event.metadata.EventMetadata
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.LocalDateTime

@Entity
@DiscriminatorValue("MoneyDepositedEvent")
class MoneyDepositedEventEntity(
    accountId: String,
    eventDate: LocalDateTime,
    amount: Double,
    metadata: EventMetadata
) : AbstractAccountEventEntity(
    id = null,
    accountId = accountId,
    eventDate = eventDate,
    amount = amount,
    metadata = metadata
) {
    companion object {
        // factory method
        fun of(
            accountId: String,
            now: LocalDateTime,
            amount: Double,
            eventMetadata: EventMetadata
        ): MoneyDepositedEventEntity {
            return MoneyDepositedEventEntity(accountId, now, amount, eventMetadata)
        }
    }
}